src/utils/process_utils/src/process_utils.cpp

```diff
@@ -43,115 +44,47 @@
+    std::string redirected_command{ command };
+    redirected_command += " ";
 
-    if (pid == 0)
+    for (const std::string& arg : args)
     {
-        // Running inside child process.
-
-        // Redirect stdout and stderr to WRITE_END
-        ADUCPAL_dup2(filedes[WRITE_END], STDOUT_FILENO);
-        ADUCPAL_dup2(filedes[WRITE_END], STDERR_FILENO);
-
-        ADUCPAL_close(filedes[READ_END]);
-        ADUCPAL_close(filedes[WRITE_END]);
-
-        std::vector<char*> argv;
-        argv.reserve(args.size() + 2);
-        argv.emplace_back(const_cast<char*>(command.c_str())); // NOLINT(cppcoreguidelines-pro-type-const-cast)
-        for (const std::string& arg : args)
-        {
-            argv.emplace_back(const_cast<char*>(arg.c_str())); // NOLINT(cppcoreguidelines-pro-type-const-cast)
-        }
-        argv.emplace_back(nullptr);
-
-        // The exec() functions only return if an error has occurred.
-        // The return value is -1, and errno is set to indicate the error.
-        int status = ADUCPAL_execvp(command.c_str(), &argv[0]);
-
-        fprintf(stderr, "execvp failed, returned %d, error %d\n", status, errno);
-
-        _exit(EXIT_FAILURE);
+        redirected_command += arg;
+        redirected_command += " ";
     }
 
-    ADUCPAL_close(filedes[WRITE_END]);
+    // We want to capture stderr as well, so we need to append "2>&1"
+    redirected_command += "2>&1";
 
-    for (;;)
+    FILE* fp = ADUCPAL_popen(redirected_command.c_str(), "r");
+    if (fp == NULL)
     {
-        char buffer[1024];
-        ssize_t count;
-        count = ADUCPAL_read(filedes[READ_END], buffer, sizeof(buffer));
-
-        if (count == -1)
-        {
-            Log_Error("Read failed, error %d", errno);
-            break;
-        }
-
-        if (count <= 0)
-        {
-            break;
-        }
-
-        // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-constant-array-index)
-        buffer[count] = 0;
-        output += buffer;
+        return errno;
     }
 
-    int wstatus;
-    int childExitStatus;
-
-    ADUCPAL_waitpid(pid, &wstatus, 0);
-
-    // Get the child process exit code.
-    if (WIFEXITED(wstatus))
-    {
-        // Child process terminated normally.
-        // e.g. by calling exit() or _exit(), or by returning from main().
-        childExitStatus = WEXITSTATUS(wstatus);
-    }
-    else if (WIFSIGNALED(wstatus))
-    {
-        // Child process terminated by a signal.
+    char buffer[1024];
 
-        // Get the number of the signal that caused the child process to terminate.
-        childExitStatus = WTERMSIG(wstatus);
-        Log_Info("Child process terminated, signal %d", childExitStatus);
-    }
-    else if (WCOREDUMP(wstatus))
+    while (fgets(buffer, sizeof(buffer), fp) != nullptr)
     {
-        // Child process produced a core dump
-        childExitStatus = WCOREDUMP(wstatus);
-        Log_Error("Child process terminated, core dump %d", childExitStatus);
+        output += buffer;
```

Comment:
need to handle calling ADUC_pclose on exception thrown.  RAII wrapper would be ideal instead of goto done.

Reply:
we didn't before -- if we're out of memory, we should probably just crash, no?  otherwise I can wrap this with a try catch.  what do you think?

Reply:
probably not worth it as an exception here indicates oom